### PR TITLE
Synced implementation of token_endpoint

### PR DESCRIPTION
**Pull request by tpazderka**
All three providers (oauth2, oic and extension) now share common code.

- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.
---

Prep work for #615 and #617.

Also getting rid of duplicated code...

### CODE REVIEW

**Comment by tpazderka** [comment]
I will work on the few missing tests tomorrow.

File: src/oic/oauth2/provider.py [hunk 1]
```diff
@@ -768,51 +769,75 @@ def token_scope_check(self, areq, info):
...
+                logger.error('Code not present in SessionDB')
+                error = TokenErrorResponse(error="unauthorized_client")
+                return Unauthorized(error.to_json(), content="application/json")
+
+            resp = self.token_scope_check(areq, _info)
+            if resp:
+                return resp
+            # If redirect_uri was in the initial authorization request verify that they match
+            if "redirect_uri" in _info and areq["redirect_uri"] != _info["redirect_uri"]:
+                logger.error('Redirect_uri mismatch')
```
Comment (schlenk) [inline src/oic/oauth2/provider.py]: Would be nice to include the two mismatched URIs in the error log message.

**Comment by tpazderka** [comment]
OK, finished the tests. There are two spots missing. They technically were not changed :) so I am not going to bother with them just yet.

**Comment by tpazderka** [comment]
Thanks for the review. I will probably do a separate issue for the refactoring of init calls. The refactoring is probably needed for most of the shared methods anyway and I would like to keep this focused on token_endpoint.

File: src/oic/oauth2/provider.py [hunk 2]
```diff
@@ -153,6 +154,8 @@ def re_authenticate(areq, authn):
 
 class Provider(object):
     endp = [AuthorizationEndpoint, TokenEndpoint]
+    # Define the message class that in token_enpdoint
+    atr_class = AccessTokenRequest
```
[the remark this answers is not in the document]
Comment (tpazderka) [reply src/oic/oauth2/provider.py]: Hm, nothing too specific. Just had the impression that the __init__ has quite a lot of kwargs already :)

**Pull request merged**